### PR TITLE
Http[Request|Response]Stream.ValidateState made inlineable

### PR DESCRIPTION
**Pull request by gfoidl**
Instead of the `switch` `if` is used and the code got tightened.
Now it is inlined in the methods, that other than the call to `ValidateState` just forward to the `Read/Write`-methods.

### CODE REVIEW

File: src/Kestrel.Core/Internal/Http/HttpResponseStream.cs [hunk 1]
```diff
@@ -145,25 +146,23 @@ public void Abort()
...
-                    throw new ObjectDisposedException(nameof(HttpResponseStream), CoreStrings.WritingToResponseBodyAfterResponseCompleted);
-                case HttpStreamState.Aborted:
-                    if (cancellationToken.IsCancellationRequested)
-                    {
-                        // Aborted state only throws on write if cancellationToken requests it
-                        cancellationToken.ThrowIfCancellationRequested();
-                    }
-                    break;
+				// Aborted state only throws on write if cancellationToken requests it
+				cancellationToken.ThrowIfCancellationRequested();
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http/HttpResponseStream.cs]: Do we know if CancellationToken.ThrowIfCancellationRequested itself gets inlined?

File: src/Kestrel.Core/Internal/Http/HttpResponseStream.cs [hunk 2]
```diff
@@ -145,25 +146,23 @@ public void Abort()
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (pakrym) [inline src/Kestrel.Core/Internal/Http/HttpResponseStream.cs]: Do you need the attribute for this method? It looks pretty small

File: src/Kestrel.Core/Internal/Http/HttpResponseStream.cs (hunk 1 above)
Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http/HttpResponseStream.cs]: Yes it gets inlined. Checked the dasm. And it's just the same property-access, so this duplication was removed. See https://github.com/dotnet/coreclr/blob/85374ceaed177f71472cc4c23c69daf7402e5048/src/System.Private.CoreLib/src/System/Threading/CancellationToken.cs#L305-L311

File: src/Kestrel.Core/Internal/Http/HttpResponseStream.cs (hunk 2 above)
Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http/HttpResponseStream.cs]: Unfortunately yes. At least on `linux-x64` where I checked the dasm.

Comment (pakrym) [reply src/Kestrel.Core/Internal/Http/HttpResponseStream.cs]: What's the reason for not inlining? Unprofitable?

**Comment by pakrym** [comment]
What motivated this change? Any before-after results?

**Comment by gfoidl** [comment]
> What motivated this change?

Stepped with the debugger through the source, spotted these methods and I knew that `switch` won't inline. Hence I submitted this PR.

> Any before-after results?

I'm just getting into the Kestrel-source, so can you give me any pointers to relevant benchmarks? Rolling out my own that cover these aspects may be not so easy (I believe).

File: src/Kestrel.Core/Internal/Http/HttpResponseStream.cs (hunk 2 above)
Comment (gfoidl) [reply src/Kestrel.Core/Internal/Http/HttpResponseStream.cs]: Yeah `unprofitable inline`.

**Comment by halter73** [comment]
I don't think we have any microbenchmarks for `Http[Request|Response]Stream` yet. Feel free to add one. For your other PR, there's the [StringUtilitesBenchmark](https://github.com/aspnet/KestrelHttpServer/blob/27799b5acb59523889a3a52a39d606e795b81c50/benchmarks/Kestrel.Performance/StringUtilitiesBenchmark.cs).

**Comment by pakrym** [comment]
You can also try https://github.com/aspnet/KestrelHttpServer/blob/dev/benchmarks/Kestrel.Performance/InMemoryTransportBenchmark.cs that runs entire pipeline in memory and see if it shows any change.

**Comment by halter73** [comment]
InMemoryTransportBenchmark is a good idea. I don't think that tests the Http*Request*Stream though.

**Comment by gfoidl** [comment]
### InMemoryTransportBenchmark

Note I ran the benchmarks with `netcoreapp2.1`, not with `netcoreapp2.0` as in the `csproj`.

#### Current dev

|             Method |     Mean |     Error |    StdDev |      Op/s |  Gen 0 | Allocated |
|------------------- |---------:|----------:|----------:|----------:|-------:|----------:|
|          Plaintext | 7.886 us | 0.1449 us | 0.1285 us | 126,807.0 |      - |     371 B |
| PlaintextPipelined | 3.914 us | 0.0775 us | 0.1060 us | 255,511.0 | 0.1221 |     432 B |

#### This PR

|             Method |     Mean |     Error |    StdDev |   Median |      Op/s |  Gen 0 | Allocated |
|------------------- |---------:|----------:|----------:|---------:|----------:|-------:|----------:|
|          Plaintext | 7.137 us | 0.1814 us | 0.5292 us | 6.869 us | 140,111.2 |      - |     368 B |
| PlaintextPipelined | 3.794 us | 0.0711 us | 0.0665 us | 3.781 us | 263,607.0 | 0.1221 |     432 B |

So it got a bit faster.
`HttpRequestStream` isn't touched by this benchmark. I assume this will also be faster, but not in the same amount as `HttpResponseStream`, due to the more complex validation.

**Comment by muratg** [comment]
@JunTaoLuo Could you take a look? Let's see if we can do this in 2.2 instead.

**Comment by JunTaoLuo** [comment]
I just re-ran the benchmarks on netcoreapp2.2 and here are the results:

Before:
```
             Method |     Mean |     Error |    StdDev |      Op/s | Allocated |
------------------- |---------:|----------:|----------:|----------:|----------:|
          Plaintext | 7.962 us | 0.1587 us | 0.3980 us | 125,598.0 |     369 B |
 PlaintextPipelined | 5.247 us | 0.0906 us | 0.0757 us | 190,575.9 |     432 B |
```
After:
```
             Method |     Mean |     Error |    StdDev |      Op/s | Allocated |
------------------- |---------:|----------:|----------:|----------:|----------:|
          Plaintext | 7.738 us | 0.1529 us | 0.2425 us | 129,240.2 |     369 B |
 PlaintextPipelined | 4.784 us | 0.0421 us | 0.0329 us | 209,020.8 |     432 B |
```

The improvements are modest. I'll merge to release/2.2 and master.

**Comment by JunTaoLuo** [comment]
Merged in https://github.com/aspnet/KestrelHttpServer/commit/773aa6e6ffa542b30165db4cde3ef3a167f28ffc. Thansk @gfoidl !

**Pull request closed without merging**